### PR TITLE
StatisticsReporter as replacement for Statistics VPP

### LINKED ISSUE #16584: Create StatisticsReporter in Stochastic Tools
## Reason

The Statistics VPP in stochastic_tools can compute various quantities including confidence intervals. However, recording the calculations in CSV format is limiting and requires numeric codes to indicate what the vectors represent. Thus, to understand the output the input file is required. 

The Reporter system allows for more generic types and JSON output, thus adds the ability to organize the calculations and compute more complex data types.

The generic, input file agnostic output is required for use with a digital twin project that requires JSON files to process with machine learning.

## Design

Convert current capability of Statistics to a StatisticsReporter and support all the current calculations.

## Impact

Statistics VPP will be deprecated.

### PR DESCRIPTION
**Pull request by aeslaughter**
(closes #16584)

### CODE REVIEW

File: modules/stochastic_tools/test/tests/reporters/statistics_reporter/gold/statistics_ci_out.json [hunk 1]
```diff
@@ -0,0 +1,186 @@
+{
+    "time_steps": [
+        {
+            "reporters": [
```
Comment (aeslaughter) [inline modules/stochastic_tools/test/tests/reporters/statistics_reporter/gold/statistics_ci_out.json:4]: @zachmprince We should talk about the design of this output, there is a lot of duplication as is.

File: modules/stochastic_tools/include/utils/Calculators.h [hunk 2]
```diff
@@ -43,67 +47,117 @@ MultiMooseEnum makeCalculatorEnum();
...
+  else if (item == "norm2")
+    return libmesh_make_unique<const L2Norm<InType, OutType>>(other, item);
+
+  else if (item == "ratio")
+    return libmesh_make_unique<const Ratio<InType, OutType>>(other, item);
+
+  ::mooseError("Failed to create Statistics::Calculator object for ", item);
+  return nullptr;
+}
+*/
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/Calculators.h:161]: @loganharbour When I try to use this template function, the compiler no like.

File: modules/stochastic_tools/src/utils/Calculators.C [hunk 3]
```diff
@@ -168,27 +146,74 @@ Ratio::compute(const std::vector<Real> & data, bool is_distributed) const
...
+
+  else if (item == "norm2")
+    return libmesh_make_unique<const L2Norm<std::vector<Real>, Real>>(other, item);
+
+  else if (item == "ratio")
+    return libmesh_make_unique<const Ratio<std::vector<Real>, Real>>(other, item);
+
+  ::mooseError("Failed to create Statistics::Calculator object for ", item);
+  return nullptr;
+}
```
Comment (aeslaughter) [inline modules/stochastic_tools/src/utils/Calculators.C:214]: This non-template version works fine. If you switch to the template version there are two or three calls that template arguments need to be added. I also attempted to set default template arguments, but that didn't help.

File: modules/stochastic_tools/include/utils/Calculators.h (hunk 2 above)
Comment (loganharbour) [reply modules/stochastic_tools/include/utils/Calculators.h:161]: Do you have a compiler error?

**Comment by aeslaughter** [comment]
@loganharbour So changing to the template version does compile, but with this warning. I didn't dig into it much yet.
```
┏ (moose) slauae@slaughter.local [statistics-reporter-16584]
┃ /Users/slauae/projects/moose/modules/stochastic_tools/test/tests/reporters
┗ $ mk
make -C /Users/slauae/projects/moose/modules/stochastic_tools  -j56
Checking if header needs updating: /Users/slauae/projects/moose/framework/include/base/MooseRevision.h...
Creating Unity /Users/slauae/projects/moose/modules/stochastic_tools/build/unity_src/vectorpostprocessors_Unity.C
Creating Unity /Users/slauae/projects/moose/modules/stochastic_tools/build/unity_src/utils_Unity.C
Compiling C++ (in devel mode) /Users/slauae/projects/moose/modules/stochastic_tools/build/unity_src/reporters_Unity.C...
Compiling C++ (in devel mode) /Users/slauae/projects/moose/modules/stochastic_tools/build/unity_src/vectorpostprocessors_Unity.C...
Compiling C++ (in devel mode) /Users/slauae/projects/moose/modules/stochastic_tools/build/unity_src/utils_Unity.C...
In file included from /Users/slauae/projects/moose/modules/stochastic_tools/build/unity_src/utils_Unity.C:2:
In file included from /Users/slauae/projects/moose/modules/stochastic_tools/src/utils/BootstrapCalculators.C:11:
/Users/slauae/projects/moose/modules/stochastic_tools/build/header_symlinks/Calculators.h:135:12: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++2a extension [-Wc++2a-extensions]
    return libmesh_make_unique<const Min<InType, OutType>>(other, item);
           ^
/Users/slauae/projects/moose/modules/stochastic_tools/build/header_symlinks/Calculators.h:138:12: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++2a extension [-Wc++2a-extensions]
    return libmesh_make_unique<const Max<InType, OutType>>(other, item);
           ^
/Users/slauae/projects/moose/modules/stochastic_tools/build/header_symlinks/Calculators.h:141:12: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++2a extension [-Wc++2a-extensions]
    return libmesh_make_unique<const Sum<InType, OutType>>(other, item);
           ^
/Users/slauae/projects/moose/modules/stochastic_tools/build/header_symlinks/Calculators.h:144:12: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++2a extension [-Wc++2a-extensions]
    return libmesh_make_unique<const Mean<InType, OutType>>(other, item);
           ^
/Users/slauae/projects/moose/modules/stochastic_tools/build/header_symlinks/Calculators.h:147:12: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++2a extension [-Wc++2a-extensions]
    return libmesh_make_unique<const StdDev<InType, OutType>>(other, item);
           ^
/Users/slauae/projects/moose/modules/stochastic_tools/build/header_symlinks/Calculators.h:150:12: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++2a extension [-Wc++2a-extensions]
    return libmesh_make_unique<const StdErr<InType, OutType>>(other, item);
           ^
/Users/slauae/projects/moose/modules/stochastic_tools/build/header_symlinks/Calculators.h:153:12: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++2a extension [-Wc++2a-extensions]
    return libmesh_make_unique<const L2Norm<InType, OutType>>(other, item);
           ^
/Users/slauae/projects/moose/modules/stochastic_tools/build/header_symlinks/Calculators.h:156:12: warning: use of function template name with no prior declaration in function call with explicit template arguments is a C++2a extension [-Wc++2a-extensions]
    return libmesh_make_unique<const Ratio<InType, OutType>>(other, item);
           ^
8 warnings generated.
Compiling C++ (in devel mode) /Users/slauae/projects/moose/framework/build/unity_src/outputs_Unity.C...
Compiling C++ (in devel mode) /Users/slauae/projects/moose/framework/src/utils/SystemInfo.C...
Compiling C++ (in devel mode) /Users/slauae/projects/moose/framework/src/base/MooseApp.C...
Linking Library /Users/slauae/projects/moose/modules/stochastic_tools/lib/libstochastic_tools-devel.la...
Linking Library /Users/slauae/projects/moose/framework/libmoose-devel.la...
Linking Executable /Users/slauae/projects/moose/modules/stochastic_tools/stochastic_tools-devel...
```

When running it does this:
```
┏ (moose) slauae@slaughter.local [statistics-reporter-16584]
┃ /Users/slauae/projects/moose/modules/stochastic_tools/test/tests/reporters/statistics_reporter
┗ $ ../../../../stochastic_tools-devel -i statistics.i

dyld: lazy symbol binding failed: Symbol not found: __ZN15StochasticTools14makeCalculatorERK13MooseEnumItemRKN7libMesh14ParallelObjectE
  Referenced from: /Users/slauae/projects/moose/modules/stochastic_tools/lib/libstochastic_tools-devel.0.dylib
  Expected in: flat namespace

dyld: Symbol not found: __ZN15StochasticTools14makeCalculatorERK13MooseEnumItemRKN7libMesh14ParallelObjectE
  Referenced from: /Users/slauae/projects/moose/modules/stochastic_tools/lib/libstochastic_tools-devel.0.dylib
  Expected in: flat namespace

zsh: abort      ../../../../stochastic_tools-devel -i statistics.i
```

I am doing something stupid, just can't find it.

**Comment by aeslaughter** [comment]
@zachmprince I would like to get this in soon, there is more work todo for Sobol calculations but I won't get to that for a while and don't want this to sit and rot. Thanks.

File: modules/stochastic_tools/test/tests/reporters/basic/tests [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  [reporters]
+    requirement = "The system shall support the calculation of statistics using"
+
+    [statistics]
+      type = JSONDiff
+      input = statistics_reporter.i
+      jsondiff = statistics_reporter_out.json
+
+      detail = "vector of data from the reporting system; include"
```
Comment (loganharbour) [inline modules/stochastic_tools/test/tests/reporters/basic/tests:33]:
```suggestion
      detail = "vector of data from the reporting system; including"
```

File: modules/stochastic_tools/test/tests/reporters/bootstrap_statistics/bca/tests [hunk 5]
```diff
@@ -0,0 +1,39 @@
+[Tests]
+  design = StatisticsReporter.md
+  issues = '#14409 #16584'
+
+  [bca]
+    requirement = "The system shall support computing bias corrected and accelerated confidence level intervals of statics"
```
Comment (loganharbour) [inline modules/stochastic_tools/test/tests/reporters/bootstrap_statistics/bca/tests:6]:
```suggestion
    requirement = "The system shall support computing bias corrected and accelerated confidence level intervals of statistics"
```

File: modules/stochastic_tools/test/tests/reporters/bootstrap_statistics/percentile/tests [hunk 6]
```diff
@@ -0,0 +1,42 @@
+[Tests]
+  design = StatisticsReporter.md
+  issues = '#14409 #16584'
+
+  [percentile]
+    requirement = "The system shall support computing percentile confidence level intervals of statics"
```
Comment (loganharbour) [inline modules/stochastic_tools/test/tests/reporters/bootstrap_statistics/percentile/tests:6]:
```suggestion
    requirement = "The system shall support computing percentile confidence level intervals of statistics"
```

File: modules/stochastic_tools/test/tests/reporters/statistics/tests [hunk 7]
```diff
@@ -0,0 +1,28 @@
+[Tests]
+  issues = '#11217 #16584'
+  design = 'StatisticsReporter.md'
+
+  [tests]
+    requirement = 'The system shall be capable of computing the statics of a data vector that'
```
Comment (loganharbour) [inline modules/stochastic_tools/test/tests/reporters/statistics/tests:6]:
```suggestion
    requirement = 'The system shall be capable of computing the statistics of a data vector that'
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
I almost feel uncomfortable with so few comments on something this large. However - the only confusing piece was the templating on the calculators and the new context. After the fishing around, it became clear.

**Comment by aeslaughter** [comment]
Thanks @loganharbour. The real need for the templates will be more apparent when I convert of the Sobol statistics calculations, which will accept a vector and return a matrix. I will also be able to specialize the bootstrap stats to operate on matrices, so we can get confidence intervals on the Sobol statistics. Also, we will be able to continue to add whatever stats we want.

**Review by loganharbour — APPROVED** [review]

**Pull request merged**